### ISSUE TITLE
Oredictionary should have a "primary name" registration

### ISSUE THREAD

**Issue by skyboy** [MinecraftForge/MinecraftForge#3083]
Retrieved with something along the lines of `String getPrimaryOreName(ItemStack)` `int getPrimaryOreId(ItemStack)` and registered with `void registerPrimaryOreName(String, ItemStack)`. 

Allowing identifying items that are _exact_ equivalents to help with ambiguity where many items are now registered with multiple unrelated names; such as the list\* entries. Primary names would be included with the standard list of ore names/ids, but have a separate singular reference as an item can not be both exactly equivalent to `logWood` and `ingotIron`.
For example, charcoal dust would have a primary name of `dustCharcoal` but standard name of `dyeBlack` and retrieving the ore names for it would return both.

This idea most likely requires further refinement from other people weighing in, however.

**Comment by ezterry** [comment]
I admit I'm confused what you are asking for.

If you have an operation that needs a (oredict) ingotIron provided, any item from the dictionary ought to be valid even if the item happens to be some strange mods "ironwood tree log" and thus also in the logWood dictionary.  What you seem to be asking for is a way for the hypothetical ironwood mod to tell your mod this item is more of a log than a iron ingot, despite it telling you its both.  What type of action would you preform differently with this information?

In short you are looking for a one (ore dict) to many (items) relationship in addition to the many to many one that already exists.  I'm just not visualizing use cases, at least not ones that will not have unexpected results.

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Comment by skyboy** [comment]
One of the primary uses is in unification of items, such as what some of the barrel and mass storage mods do, or other mods which more explicitly over a 'converter' block. Occasionally mods register things for use in recipes that are not exact equivalents, and shouldn't be unified - one of the best examples of this, is the `listall*` entries, such as `listallWater` where you have water buckets, water bottles, and in a few mods, a stackable water item crafted from a water container which comes in a larger amount than 1. In the latter case in particular, you do not want to unify the water items into water buckets.
Not all of the use-cases are as clearcut as "blacklist `.startsWith("listall")`" however. This particular feature would alleviate that particular problem. There's also potential for other uses that do not come to mind, as I obviously cant think of everything.

**Comment by ezterry** [comment]
certainly as implemented ore dict is bad for merging near identical versions of items. In the "ironwood" you probably don't want it to merge.. but in two mods copper ingots you probably do.

For merging into storage this almost seems a place for a 'get merge type' capability. If both the source/target item stack have the merge type capability and return the same name (string) for it, the mod can merge them. (or if you want more flexibility two functions one to get the type and one to say if it can merge or not).

This could I guess be done by adding a 'merge' tag to the ore dict or convention that 'mergeable<type>' be for mergeable items but seems like forcing the oredict to do something its not intended for. Particularly since all these methods require the modder to add in the logic for the relevant items.

**Comment by cpw** [comment]
Unification use cases are really pack dependent. A 'master' entry, as you ask, should be strictly a configuration element, not programmatic, since no mod should depend on a preferred choice- that breaks the whole idea of ore dictionary. To be clear, you should be able to query for the preferred item, but setting it would be configuration only. (And some default, probably the first in class, in absence of configuration)

**Comment by skyboy** [comment]
Not `<String, ItemStack>`, @cpw; `<ItemStack, String>` vs `<ItemStack, List<String>>`.

There would be no globally preferred item suggested or provided by forge, just the ability to say "These two copper ingots are exactly the same" and "These two water items are merely equivalent, not exactly the same." What item to unify into is an implementation detail for the mods to solve, this would merely provide the facility to do so with decreased occurrences of resource duplication and transmutation.

**Comment by RANKSHANK** [comment]
Liquid quantification is already implemented; the fluid capability solves that by providing access to the exact volume of the contained fluid stacks.

I guess you could have something like a capability that provides a material stack which would be standardized to milliblocks (eg iron bars consisting of 112 milliblocks of iron)... but the overhead to payoff ratio for that is questionable at best(Concept is pretty tantalizing though). Aside from that it's impossible to tell how much of a given material a modder provided string is worth.

**Comment by skyboy** [comment]
It's not about liquids in particular, milk buckets aren't a standard part of the fluid dict, but they're subjected to this in a the same manner as water, and you can have the same issue applying to entirely non-fluid items like meats (`listallMeat` being most obvious to me): you don't exactly want to unify zombie flesh into cooked steak.

**Comment by RANKSHANK** [comment]
And forge has inbuilt support for treating milk as a liquid on request.

So how exaclty do you plan on implementing a standard of specifity then? Sounds like an unyieldy tree of correlations that people would need to muck around with until they get that 'just right' level of abstraction on a given material. IE meat. Maybe some dude's making hotdogs and doesn't care what meat goes into them. Can't make everyone happy.

**Comment by skyboy** [comment]
Details of how a mod determines what materials are suitable for crafting results are beyond scope of this issue and would most likely use standard oredict methods as it does now; or a custom recipe testing if a wolf can eat the food item. This issue is about helping to deal with the many mass storage methods that want to store a type of item (e.g., `ingotCopper`) for all suitable items, but not incidentally also store items that are merely crafting equivalents (e.g., `listallWater`).

**Comment by RANKSHANK** [comment]
I'm 'fixating' on a system that is proven and working, and I have tied it in to a concept that follows said system's structure. Seeing as the primary usage of materials is through some form of crafting, that facet seems like a fairly valid subject for discussion. 

My point of view is that this should be skewing more towards something with a basis in numbers as opposed to interpretation of a string. With the buckets, what is the bucket comprised of? You don't really want to have your diamond buckets transformed into iron buckets because they're listed with a primary identifier of bucketWater. How do you check a string to see if the given ItemStack's composition matches *enough* to be considered equivalent to another ItemStack? In fact with a lot of relatively abstract registry names; ie shardGlass, dustCharcoal, etc., at what point can you assume equivalency? How much charcoal is in charcoal dust? How big is a glass shard?
Most importantly: how do you choose which mod's Item instance in that registry is the correct return value for all other entries? The String registry matches the crafting format for the most part, but with actual stacking of multiple item instances it all falls apart. There's too much ambiguity in the dictionary to assume equivalency for direct conversion of items. The ore dictionary is great for instances where an item can say "hey I can be used as a gold ingot", but is horrid for instances of items saying "I'm exactly the same as that."
As I brought up before, something like a material capability where you can say that an instance of IronWood consists of 112 milliblocks of iron and 888 milliblocks of wood gives absolute composition which is by far more useful for the prospect of stacking/storing anonymous items together than assumptions wrought from an ambiguous string.

TLDR; Equivalency would take a different registry approach than matching utilization cases in crafting.

**Comment by williewillus** [comment]
^ that is a nice idea, but I feel like that is straying wayyyy beyond what Forge is meant for.
Forge is not EE3 or ProjectE, it should not be micro-managing material values and what's-made-of-what values.

**Comment by RANKSHANK** [comment]
@williewillus Definitely agree with it being outside of the scope of Forge :P

I was just providing a semi tangible solution instead of just flat out stating my opinion that providing means for modders to assume items are directly equivalent based on a string representation of what it is primarily used for is a horrible idea.

**Comment by skyboy** [comment]
@RANKSHANK This is explicitly for adding a feature that modders would be aware (via documentation on the method) is explicitly for the exact case you're describing as being ambiguous.
The point at which exact equivalency is assumed is *what this issue is about*. The methods described here are for determining *exact equivalency* between items like charcoal dusts, copper ingots, and so forth. Your iron wood example would not be registered in the proposed API at all, because it *isn't* something multiple mods are adding as a standard material, such as copper, tin, steel, iron nuggets, and so on. There's even a radically different type of iron wood in an existing mod, and it would be up to you to not generally screw up the registrations.
In the case of your diamond bucket example, mods wouldn't register water buckets as a primary name for anything, because they're not exactly the same. Any mod registering these entries would be doing so in error, and be causing material transmutation.

And your proven working system, while working and proven, is entirely unrelated to the issue at hand. I'll keep repeating myself for you: This issue is for determining exactly equivalent items and then having a mod in an implementation specific manner unify these items for things like mass storage: barrels, drawers, etc. It solves the issue of the 5 to 15 copper ingots you may encounter failing to all go into the same storage device.
While stacking issues persist, those are beyond the scope of forge, and if this API was present another mod could implement the behavior that lets you merge two different item stacks if they are exact equivalents.
